libpromises/policy.c

```diff
@@ -2627,6 +2702,63 @@ static bool CheckScalarNotEmptyVarRef(const char *scalar)
     return (strcmp("$()", scalar) != 0) && (strcmp("${}", scalar) != 0);
 }
 
+static bool ValidateCustomPromise(const Promise *pp, Seq *errors)
+{
+    assert(pp != NULL);
+
+    const char *const promise_type = pp->parent_section->promise_type;
+    bool valid = true;
+    Seq *attributes = pp->conlist;
+    const size_t length = SeqLength(attributes);
+    for (size_t i = 0; i < length; ++i)
+    {
+        Constraint *attribute = SeqAt(attributes, i);
+        const char *name = attribute->lval;
+        if (StringEqual(name, "action")
+            || StringEqual(name, "action_policy")
+            || StringEqual(name, "ifelapsed")
+            || StringEqual(name, "expireafter")
+            || StringEqual(name, "log_level")
+            || StringEqual(name, "classes")
+            || StringEqual(name, "comment")
+            || StringEqual(name, "depends_on")
+            || StringEqual(name, "handle")
+            || StringEqual(name, "if")
+            || StringEqual(name, "ifvarclass")
+            || StringEqual(name, "meta")
+            || StringEqual(name, "unless")
+            || StringEqual(name, "with"))
```

Comment:
Man, `pp->conlist` should really be a hashtable...

Reply:
Yep.